### PR TITLE
feat: change XTX & DLQ lifecycle to adaptive timeouts

### PR DESCRIPTION
**Pull request by MaciejBaj**
# Summary of changes

**_Summarize_** the description of your changes.

If it fixes a bug or creates a feature, link the issue.

# Acceptance Checklist

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

## Dependencies

- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any teams which manage the dependencies have been notified of the breaking changes

## Bug (non-breaking change which fixes an issue):

- [ ] Have you **_written tests_** to protect against future occurrences of the bug?

## Feature (non-breaking change which adds functionality)

- [ ] Have you **_seen it working_** in a development environment?
- [ ] Have you **_written tests_** for **_happy_** and **_unhappy_** paths?
- [ ] Have you implemented this feature as per the **_issue acceptance criteria_**?

### Substrate

If this change involves substrate, have you remembered:

- [ ] Storage Migration?
- [ ] RPC methods?
- [ ] Chainspec, both JSON specs & the module?
- [ ] Runtime versioning?
- [ ] Benchmarks?

## Breaking change (a feature that would cause existing functionality not to work as expected)

- [ ] Is the breaking change **_documented_**?
- [ ] Are your commits fully representative of the change?
- [ ] Has any previous functionality been **_deprecated_** and versioned?

## CI/CD

- [ ] If introducing new actions, have you **_looked at the action code_** for anything spurious?
- [ ] Have you written **_custom scripts_** for things that could be actions already on the marketplace?
- [ ] If introducing new actions, have you **_pegged a static version_**?

## Documentation Update

- [ ] Have you checked other documentation, such as the docs repository?
- [ ] If updating script documentation, have you **_tested it on both environments_**?

## Further comments

Feel free to explain in further detail your choices if this is a significant change.

### Summary by OpenAI

Release Notes:

- New Feature: Introduced `adaptive_timeout` variable and modified related method calls, data structures, and test cases in the Circuit module.
- Bug fix: Replaced a function call from `Circuit::process_revert_xtx_queue` to `Circuit::process_emergency_revert_xtx_queue` in multiple modules.
- New Feature: Added new functions and structs to the `Xdns` trait and removed an import statement for `TargetId`.
- Refactor: Modified code related to processing revert transactions and updated heartbeat values in various modules.
- Refactor: Changed the `CircuitDLQ` associated type in the `pallet_xdns::Config` trait to use the `Circuit` type.
- Refactor: Modified arguments and circuit roles in the `Machine` module.
- New Feature: Extended the `Portal` trait with a new function.
- Refactor: Replaced a function call in the `impl pallet_clock::traits::OnHookQueues` and `impl pallet_xdns::Config` implementations.

> "Code changes dance,
> Bugs have no chance.
> Features bloom and grow,
> With each line we sow.
> 🎉 Celebrate this PR,
> For it takes us far!"

### CODE REVIEW

File: pallets/xdns/src/lib.rs [hunk 1]
```diff
@@ -769,6 +754,51 @@ pub mod pallet {
...
+            verifier: &GatewayVendor,
+            epoch_duration_in_remote_blocks: T::BlockNumber,
+            epoch_duration_in_local_blocks: T::BlockNumber,
+        ) {
+            const N: u32 = 10; // Number of epochs to consider for the moving average
+
+            let mut history = EpochHistory::<T>::get(verifier).unwrap_or_default();
+            let history_len = history.len() as u32;
+
+            if history_len >= N {
```
Comment (palozano) [inline pallets/xdns/src/lib.rs:768]: How about the case where `history_len` is (way) larger than `N`? By making `history.remove(0)`, is it enough? Or should be something like `history.remove(history.len() as u32 - N)`?

File: pallets/circuit/vacuum/src/lib.rs [hunk 2]
```diff
@@ -542,13 +546,44 @@ mod tests {
...
             // Trigger XTX revert queue and expect move to DLQ
-            Circuit::process_revert_xtx_queue(
+            Circuit::process_emergency_revert_xtx_queue(
                 System::block_number(),
                 System::block_number(),
                 u64::MAX,
             );
             // Verify that XTX is in DLQ
-            assert_eq!(Circuit::get_dlq(xtx_id), Some(System::block_number()));
+            assert_eq!(
```
Comment (palozano) [inline pallets/circuit/vacuum/src/lib.rs:555]: These `assert_ok!` are fine if panicking?

**Review by palozano — APPROVED** [review]
Just a couple of comments.

Also, build fails because there's a timeout as an integer somewhere instead of the adaptative struct.


```
Error[E0308]: mismatched types
Error:     --> pallets/attesters/src/lib.rs:2164:30
     |
2164 |                 timeouts_at: 1,
     |                              ^ expected struct `AdaptiveTimeout`, found integer
     |
     = note: expected struct `AdaptiveTimeout<_, [u8; 4]>`
                  found type `{integer}`
```

File: pallets/xdns/src/lib.rs (hunk 1 above)
Comment (MaciejBaj) [reply pallets/xdns/src/lib.rs:750]: that's a good comment actually 👍

File: runtime/mini-mock/src/hooks.rs [hunk 3]
```diff
@@ -1,30 +1,59 @@
...
             > 100
         {
             log::error!(
                 "GlobalOnInitQueues::Invalid shares exceed 100%, returning 0 - re-check the shares"
             );
             return 0
         }
+        const BLOCKS_PER_WEEK: BlockNumber = 7 * 24 * 60 * 60;
+        const BLOCKS_PER_DAY: BlockNumber = 24 * 60 * 60;
+        const BLOCKS_PER_HOUR: BlockNumber = 60 * 60;
```
Comment (palozano) [inline runtime/mini-mock/src/hooks.rs:37]:
```rust
const BLOCKS_PER_HOUR: BlockNumber = 60 * 60;
const BLOCKS_PER_DAY: BlockNumber = BLOCKS_PER_HOUR * 24;
const BLOCKS_PER_WEEK: BlockNumber = BLOCKS_PER_DAY * 7;
```

File: runtime/mini-mock/src/hooks.rs [hunk 4]
```diff
@@ -1,30 +1,59 @@
...
+
+        if (n % BLOCKS_PER_HOUR).is_zero() {
+            let hourly_weight_limit: Weight =
+                Perbill::from_percent(HOURLY_SHARE).mul_ceil(on_init_weight_limit);
```
Comment (palozano) [inline runtime/mini-mock/src/hooks.rs:41]: Do we care to know if it saturates when multiplying?

File: runtime/mini-mock/src/hooks.rs [hunk 5]
```diff
@@ -94,4 +123,32 @@ impl pallet_clock::traits::OnHookQueues<Runtime> for GlobalOnInitQueues {
 
         total_consumed
     }
+
+    fn process_weekly(_n: BlockNumber, _hook_weight_limit: Weight) -> Weight {
+        0
```
Comment (palozano) [inline runtime/mini-mock/src/hooks.rs:128]: Is this unimplemented!()?

File: runtime/mock/src/circuit_config.rs [hunk 6]
```diff
@@ -36,26 +36,56 @@ impl t3rn_primitives::EscrowTrait<Runtime> for Runtime {
...
             return 0
         }
+
+        const BLOCKS_PER_WEEK: BlockNumber = 7 * 24 * 60 * 60;
+        const BLOCKS_PER_DAY: BlockNumber = 24 * 60 * 60;
+        const BLOCKS_PER_HOUR: BlockNumber = 60 * 60;
+
+        if (n % BLOCKS_PER_HOUR).is_zero() {
+            let hourly_weight_limit: Weight =
+                Perbill::from_percent(HOURLY_SHARE).mul_ceil(on_init_weight_limit);
```
Comment (palozano) [inline runtime/mock/src/circuit_config.rs:73]: Same as before

**Review by palozano — APPROVED** [review]
I don't think I've missed something too bad. I left a couple of comments around. Also, needs to be updated with respect Development.

File: runtime/mini-mock/src/hooks.rs (hunk 4 above)
Comment (MaciejBaj) [reply runtime/mini-mock/src/hooks.rs:41]: na it taking percentage should be less than weight limit, saturating is complex here due to types: saturating_mul function of the Perbill type expects another Perbill as an argument, but on_init_weight_limit is a u64 in your code.

File: runtime/mini-mock/src/hooks.rs (hunk 5 above)
Comment (MaciejBaj) [reply runtime/mini-mock/src/hooks.rs:128]: for now yes, I'm gonna drop queues from attesters and rewards here in next PRs

**Pull request merged**